[gh release create/upload] Expand glob patterns for all platforms

iamazeem opened the pull request:
Fixes #5099.

Verified that this issue is **only applicable to Windows Command Prompt and Powershell**.
See this relevant [SO thread](https://stackoverflow.com/questions/72434739/globbing-patterns-in-windows-command-prompt-powershell) for more details.

GitBash behaves same as a Linux shell.

This issue affects `gh release create` and `gh release upload` commands.
In this PR, the glob patterns are expanded internally for Windows.

### Tests (same for Windows Command Prompt and Powershell)

```powershell
# Create test files e.g. test.{1,2,3} and log.{1,2,3}
> echo "test.1" > test.1
> echo "test.2" > test.2
> echo "test.3" > test.3
> echo "log.1" > log.1
> echo "log.2" > log.2
> echo "log.3" > log.3

# Create release with test.{1,2,3} files with glob pattern test.*
> .\gh.exe release create --repo iamazeem/test --draft --notes "Test" v1.2.3 test.*
https://github.com/iamazeem/test/releases/tag/untagged-d72acf4185040741cacb

# Verify release
> .\gh.exe release view --repo iamazeem/test v1.2.3
v1.2.3
Draft • iamazeem created this less than a minute ago

  Test


Assets
test.1  18 B
test.2  18 B
test.3  18 B

View on GitHub: https://github.com/iamazeem/test/releases/tag/untagged-d72acf4185040741cacb


# Upload log.{1,2,3} files to release with glob pattern log.*
> .\gh.exe release upload --repo iamazeem/test v1.2.3 log.*
Successfully uploaded 3 assets to v1.2.3

# Verify release again
> .\gh.exe release view --repo iamazeem/test v1.2.3
v1.2.3
Draft • iamazeem created this about 2 minutes ago

  Test


Assets
log.1   16 B
log.2   16 B
log.3   16 B
test.1  18 B
test.2  18 B
test.3  18 B

View on GitHub: https://github.com/iamazeem/test/releases/tag/untagged-d72acf4185040741cacb
```

#### Invalid patterns

```powershell
# Create release with a glob pattern which doesn't expand e.g. test123.*
> .\gh.exe release create --repo iamazeem/test --draft --notes "Test" v1.2.3 test123.*
CreateFile test123.*: The filename, directory name, or volume label syntax is incorrect.
```

**Notes**

- In **Powershell**, quoting (single or double) glob patterns work the same and the quotes are not passed to `gh`.
- In **Command Prompt**, double quotes work same as Powershell but single quotes become part of the pattern itself.
  - For example, `'test.*'` is passed as-is and cannot be expanded.

**Error on Command Prompt**

```cmd
> .\gh.exe release create --repo iamazeem/test --draft --notes "Test" v1.2.3 'test.*'
CreateFile 'test.*': The filename, directory name, or volume label syntax is incorrect.
```
---

## UPDATE

After discussions, the support for glob patterns has been added for all the platforms.

iamazeem committed:
[gh release create/upload] Expand glob patterns on Windows

jtmcg commented on pkg/cmd/release/shared/upload.go:
```diff
@@ -63,6 +72,22 @@ func AssetsFromArgs(args []string) (assets []*AssetForUpload, err error) {
+		}
+		if len(matches) > 0 {
+			assets = append(assets, matches...)
+		} else {
+			assets = append(assets, pattern)
+		}
```
Can you explain the motivation behind this `else` block? It seems to me that appending the unmatched pattern to the assets we're trying to find would still cause the same problem on Windows that currently exists. What exactly is the behavior, here?

iamazeem replied:
An invalid pattern won't be expanded and it's an input issue.
The `else` block is here to keep the behavior consistent w.r.t. other OSes e.g. Linux.

Here's an example from **Linux** (Ubuntu):

```shell
$ bin/gh release create --draft --notes "Test" v1.2.3 test.* test123.*
stat test123.*: no such file or directory
```

`test.*` is a valid glob pattern but `test123.*` is not because no expansion exists for it hence the error.

Tried to keep the behavior consistent on Windows as well:

```cmd
> .\gh.exe release create --repo iamazeem/test --draft --notes "Test" v1.2.3 test.* test123.*
CreateFile test123.*: The filename, directory name, or volume label syntax is incorrect.
```

jtmcg reviewed (changes requested):
Coming in with another blast from the past! We appreciate all the work you're putting in, @iamazeem 🙌 

I reviewed the old issue and I think there was just a bit more to do from there than is delivered in this PR. I've added [Acceptance Criteria](https://github.com/cli/cli/issues/5099#issuecomment-2648624125) to it so we can more easily track what needs doing. Let me know if you have any questions.

**tl;dr:** We want to expand wildcards for ALL platforms using `gh` and we want this to happen in `gist create` and `release create` in addition to what you've done here for `release upload`. I suspect its not too much more effort for those, given the solution you've outlined here 🙂 

Also, if you haven't already come to expect it from me by now, I'm going to ask for some tests 😅

iamazeem commented:
@jtmcg:

> We want to expand wildcards for ALL platforms using `gh`

As mentioned in the PR description, it's only needed for Windows.
The shells on Linux and Mac perform their own expansion.
That's why I targeted only Windows after verification.

> we want this to happen in `gist create`

Sure, I'll add its support for `gist create`.

> Also, if you haven't already come to expect it from me by now, I'm going to ask for some tests 😅

Actually, I thought about how to approach these tests and then decided not to.
Apparently, it's a shell-specific thing and didn't feel like that it should be covered.
If you could share some examples that'd be very helpful. Thanks!

iamazeem committed:
Fix `gh gist create` for multiple filenames and glob patterns

iamazeem commented:
## `gh gist create`

Working fine both on Linux and Windows.

### Tests (Windows)

```cmd
> .\gh.exe gist create -d "test" test.*
- Creating gist with multiple files
✓ Created secret gist test.1
https://gist.github.com/iamazeem/4d5a636d1c66f58be21440efd37247f0

>.\gh.exe gist list -L1
ID                                DESCRIPTION  FILES    VISIBILITY  UPDATED
4d5a636d1c66f58be21440efd37247f0  test         3 files  secret      less than a minute ago

> .\gh.exe gist view 4d5a636d1c66f58be21440efd37247f0
test

test.1

test.1

test.2

test.1

test.3

test.1

> .\gh.exe gist delete --yes 4d5a636d1c66f58be21440efd37247f0
✓ Gist "test.1" deleted
```

iamazeem commented:
@jtmcg: Please review. Thanks!

iamazeem committed:
Merge branch 'trunk' into 5099-gh-release-create-upload-expand-glob-patterns-on-windows

SahibYar commented on pkg/cmdutil/args.go:
```diff
@@ -57,3 +59,22 @@ func NoArgsQuoteReminder(cmd *cobra.Command, args []string) error {
+	if runtime.GOOS != "windows" {
+		return patterns, nil
+	}
+	var expansions []string
```
Avoid unnecessary reallocations.
```
expansions := make([]string, 0, len(patterns)) // Preallocate slice
```

iamazeem replied:
Seems like a GenAI suggestion. :thinking: 

Given the context, it seems like a premature optimization. How do you see it'll help?
Some benchmarks would be helpful to understand its impact for different user inputs.

jtmcg replied:
Could be, though it's not a bad suggestion... It is an optimization with tiny returns that probably don't give meaningful gains to performance in this situation, but I'm generally for building these habits and appreciate suggestions/callouts like this, especially since I come from a Python/JavaScript background that hand-waves away memory allocation and pointers 😅

jtmcg replied:
That said, there's no guarantee that the number of expansions match the number of patterns, so this pre-allocation of memory doesn't actually work...

iamazeem replied:
> That said, there's no guarantee that the number of expansions match the number of patterns, so this pre-allocation of memory doesn't actually work...

@jtmcg: You got it! :+1: :trophy: :smile: 
Working with C, C++ and other low-level code, I've learnt that it rarely is that simple.
Now, I always take such statements/suggestions about optimization with a grain of salt.
I always follow the "**measure first and optimize later**" approach; and, the context matters, a lot.
By context here, I mean: compiler, compiler flags, linker, target OS, hardware, CPU cores, RAM, cache, vector instructions, and everything else that directly or indirectly affects a hotpath.
IMO, without proper profiling/benchmarking given a certain context, such optimizations and their discussions are generally futile. :sweat_smile:

jtmcg replied:
💯

SahibYar commented on pkg/cmdutil/args.go:
```diff
@@ -57,3 +59,22 @@ func NoArgsQuoteReminder(cmd *cobra.Command, args []string) error {
+	for _, pattern := range patterns {
+		matches, err := filepath.Glob(pattern)
+		if err != nil {
+			return nil, fmt.Errorf("%s: %s", pattern, err)
```
```
fmt.Errorf("%s: %w", pattern, err) // Use %w for wrapped errors
```
Allows error wrapping, making debugging easier.

jtmcg commented:
> As mentioned in the PR description, it's only needed for Windows.
> The shells on Linux and Mac perform their own expansion.
> That's why I targeted only Windows after verification.

Totally. There may be a good reason to also handle other os's explicitly in the CLI, though... Not sure what they are at the moment, but I'll think on this and get back to you on it 👍 If you can think of any reasons for or against it, I'm all ears, too

> Actually, I thought about how to approach these tests and then decided not to.
> Apparently, it's a shell-specific thing and didn't feel like that it should be covered.
> If you could share some examples that'd be very helpful. Thanks!

Sure thing! I'll take a look and add a PR into your feature branch for it if I can find a solution

jtmcg commented:
@iamazeem, here's the [PR](https://github.com/iamazeem/cli/pull/2) with the testing example. Note, I moved the os check out of the helper function and into the commands. I'm not sure if that's exactly right, but it seemed weird to have a function named `GlobWindowsPaths` that was also making decisions about `mac` and `ubuntu` 🤷

iamazeem commented:
@jtmcg

> There may be a good reason to also handle other os's explicitly in the CLI, though... Not sure what they are at the moment, but I'll think on this and get back to you on it 👍 If you can think of any reasons for or against it, I'm all ears, too

On nix shells, the expansions happen when args are in double quotes e.g. `"test.*"`.
In single quotes, the glob patterns are not expanded e.g. `'test.*'`.
If `gh` were to start expanding everything internally then the behavior of single quotes would be altered.
That was my only concern while implementing this i.e. not to break the default behavior of nix shell expansions and the UX should remain consistent on all platforms where `gh` works.

jtmcg committed:
Add testing to GlobWindowsPaths (#2)

jtmcg commented on pkg/cmdutil/args.go:
```diff
@@ -57,3 +58,21 @@ func NoArgsQuoteReminder(cmd *cobra.Command, args []string) error {
+		if len(matches) > 0 {
+			expansions = append(expansions, matches...)
+		} else {
+			expansions = append(expansions, pattern)
```
Alright, I talked to @andyfeller about this and we think this should be erroring here. Additionally, we think that ALL systems should rely on this Glob expansion and not just Windows. Here's why:

Currently, the desired behavior is that we error if no matches are found for a pattern. Right now, this is happening [here](https://github.com/cli/cli/blob/537a22228cd6b42b740d7f1c09f47c45bb1dab30/pkg/cmd/release/shared/upload.go#L47-L51) in `AssetsForArgs`. The issue is that its not really the same error for different filesystems. The error ALWAYS happens for Windows, because it doesn't auto-expand glob patterns, and we find that misleading.

Thus, to reflect the behavior of shells, we want the CLI to error when it finds no expansions. That way we can separate out the "this didn't match any files" errors from other errors. If we don't do that, then we are essentially splitting the error handling for no matches between two places that is shell dependent, and we don't really like that pattern.

I'll write a high-level decision of the desired behavior in another comment.

iamazeem replied:
Right. Makes sense. I'll update the PR accordingly. :+1:

andyfeller replied:
Going to retroactively capture my thoughts as these changes are being discussed when cutting a new release.

- I appreciate some shells will expand patterns before calling executables
- I dislike engineering around OS and shell specific behaviors like this from a product perspective as it requires broader and deeper insight than might be apparent to the maintainer
- I can live with us limiting this behavior strictly to Windows OS until we hear about other situations and revisit the decision

jtmcg reviewed (commented):
Great, so here's where we're at:

1. We want to handle all globbing ourselves, regardless of shell
2. We want to error if a glob pattern doesn't find any matches. This error should contain ALL patterns that don't have matches and not just the first that didn't match. Perhaps this suggests a new bespoke error type?

Does this make sense?

iamazeem committed:
Support globbing for all platforms

jtmcg commented on pkg/cmd/gist/create/create.go:
```diff
@@ -85,7 +88,7 @@ func NewCmdCreate(f *cmdutil.Factory, runF func(*CreateOptions) error) *cobra.Co
 		Aliases: []string{"new"},
 		RunE: func(c *cobra.Command, args []string) error {
-			opts.Filenames = args
+			opts.Filenames = args[:]
```
Curious. What's your reasoning for slicing, here?

iamazeem replied:
Removed with https://github.com/cli/cli/pull/10413/commits/8ace163608091804a42bd652b4b8bd7da3836f06. Not needed.

jtmcg commented on pkg/cmdutil/args.go:
```diff
@@ -57,3 +59,24 @@ func NoArgsQuoteReminder(cmd *cobra.Command, args []string) error {
+	expansions := []string{}
+
+	for _, pattern := range patterns {
+		if pattern == "-" || strings.Contains(pattern, "#") {
+			expansions = append(expansions, pattern)
```
What's this for?

Also, if we want to keep it, we should cover this path with tests

iamazeem replied:
This is to ignore the expansion when `-` is part of the args.
And, the `#` needs to be checked for args for filenames with labels i.e. `filename#label`.
Both code paths are already covered in other tests in `gist` and `release`.
You may comment this part and observe the tests locally.

iamazeem replied:
Added tests for `-` and `filename#label` cases with https://github.com/cli/cli/pull/10413/commits/3d726c9865ae545ddc0802986c5f93187be36812.

jtmcg replied:
Ah, cool. I guess I'm happy that the tests in `gist` and `release` showed you this was needed, here. That said, I think we're crossing the streams again. I don't think that `GlobPaths` should care about these special use-cases that are surfaced in `gist` and `release`. They should be handled where they matter.

- In the case of `gists`, I believe that means we should handle them near the top of [createRun](https://github.com/cli/cli/blob/3d726c9865ae545ddc0802986c5f93187be36812/pkg/cmd/gist/create/create.go#L108-L115)
- In the case of `release`, it seems that the functionality is [sort of already there in AssetsFromArgs](https://github.com/cli/cli/blob/3d726c9865ae545ddc0802986c5f93187be36812/pkg/cmd/release/shared/upload.go#L47) and we should keep it there.

In other words, I don't want `GlobPaths` to do anything but glob. Any special cases that aren't globbing should be handled where they matter.

iamazeem replied:
Went through all the potential change points.
The STDIN `-` case is generic which may be handled by `GlobPaths` internally.
For `filename#label` case, and other similar cases (maybe in the future), it seems more appropriate to handle through an exclusion function parameter (`exclude` callback).
With this approach, the code remains clean and readable.
The alternative seemed somewhat messy and not as straightforward.
Please review https://github.com/cli/cli/pull/10413/commits/f4b65b785e9f0025b0e674cf78050a115924128d.

jtmcg replied:
This still feels to interdependent to me... Why should globbing care about special syntax that `gh gist create` and `gh release create` respect?

I've opened a [PR](https://github.com/iamazeem/cli/pull/3) against this branch to demonstrate what I mean. I've also asked the team to take a look at it, because maybe this is just me being too nit picky 😅 Thanks for your patience!

iamazeem replied:
No worries. :smile: 
Let's wait for the team. :+1: 

Regarding your PR, initially I reached a similar solution, sans generics.
I'm more inclined towards a consolidated function e.g. `GlobPathsWithExclusions` or something similar that accepts an `exclude` callback, in addition to `GlobPaths`.
The `Paths` may be dropped i.e. `cmdutil.Glob` and `cmdutil.GlobWithExclusions`. :thinking:

jtmcg replied:
I value the composability of not using the consolidated function. That, and passing in callbacks as args can have a tendency to obscure nuances in situations like these that are easier to parse with in-line logic. Plus, testing is WAY easier 😅 

Also, the team took a look and are on board with the approach in the PR. However, @williammartin did call out that maybe `cmdutil` isn't the right place for this. He suggested a new package, but let's merge my PR into this first before having that discussion.

jtmcg reviewed (commented):
Looking good. I like how clean this reads, now! Have a couple of questions, though

iamazeem committed:
Remove slicing

iamazeem committed:
Add tests for - and label args

jtmcg reviewed (changes requested):
The requested changes are described in this [comment](https://github.com/cli/cli/pull/10413#discussion_r1970470576)

iamazeem committed:
Add `exclude` callback function

iamazeem committed:
Merge branch 'trunk' into 5099-gh-release-create-upload-expand-glob-patterns-on-windows

jtmcg committed:
Separate partitioning from globbing in cmdutil/args package and consumers (#3)

* Separate partitioning from globbing in cmdutil/args package and consumers

In the previous commit, GlobPaths was overloaded, containing logic
specific to command use-cases. This commit removes that functionality from
GlobPaths and back into the commands that have the special use-cases.

To do this, I've introduced a new Partition util in cmdutil/args.go that
will separate a slice into two slices given a predicate. This functionality
is leveraged by both the special use-cases described above to separate the
command-specific syntax from the globable filepaths.

* Add test to validate that the order of '-' in gh gist create args doesn't matter

darylgwpo commented:
[Audio-JACK2-FFI-main.zip](https://github.com/user-attachments/files/19374784/Audio-JACK2-FFI-main.zip)

qoqnoos1 commented:
iamazeem:5099-gh-release-create-upload-expand-glob-patterns-on-windows
